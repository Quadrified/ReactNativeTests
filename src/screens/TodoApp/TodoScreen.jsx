import React, { useState } from 'react';
import { View, StyleSheet, Pressable, Keyboard, Text } from 'react-native';
import Ionicons from 'react-native-vector-icons/Ionicons';
import AddTodo from '../../components/AddTodo';
import TodoList from '../../components/TodoList';
import AppColors from '../../themes/AppColors';

export default function TodoScreen({ navigation }) {
  const [listInput, setListInput] = useState('');
  const [list, setList] = useState([]);
  const [error, setError] = useState(false);

  const handleInput = text => {
    setListInput(text);
  };

  const onAddTodo = () => {
    if (!listInput) {
      setError(true);
      console.log('No input');
      return;
    }

    const newTodo = {
      id: Math.random() * 1000,
      value: listInput,
    };

    setList([newTodo, ...list]);
    setListInput('');
    setError(false);

    console.log(`Added todo => ${listInput}`);
  };

  const handleDelete = id => {
    console.log(
      `Deleted todo item => ${list.find(item => item.id === id).value}`,
    );
    setList(list.filter(item => item.id !== id));
  };

  const onDismissKeyboard = () => Keyboard.dismiss();

  const goToRecipesApp = () => navigation.navigate('Login');

  return (
    <Pressable onPress={onDismissKeyboard} style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.headerText}>Todo List</Text>
        <Pressable role="RecipesLoginButton" onPress={goToRecipesApp}>
          <Ionicons name="fast-food-sharp" size={24} color={AppColors.white} />
        </Pressable>
      </View>
      <View style={styles.inputArea}>
        <AddTodo
          onAddTodo={onAddTodo}
          value={listInput}
          onChange={handleInput}
          error={error}
        />
      </View>
      <View style={styles.listArea}>
        <TodoList data={list} onDelete={handleDelete} />
      </View>
    </Pressable>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: AppColors.background,
  },
  header: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    backgroundColor: AppColors.primary,
    padding: 15,
  },
  headerText: {
    color: AppColors.white,
    fontSize: 24,
  },
  inputArea: {
    marginVertical: 16,
    marginHorizontal: 10,
  },
  listArea: {
    flex: 1,
    paddingHorizontal: 15,
    alignItems: 'center',
    marginVertical: 15,
  },
});
