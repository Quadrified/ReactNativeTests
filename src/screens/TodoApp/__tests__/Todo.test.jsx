/**
 * 1. Test Cases
 */
// Create Todo item

// Create Multiple Todo Items

// Delete Todo Item

// Show error message on empty input

// Error message should disappear once a valid item is created

/**
 * 2. Import stuff
 */
import React from 'react';
import { it, expect, describe } from '@jest/globals';
import { render, fireEvent } from '@testing-library/react-native';
import TodoScreen from '../TodoScreen';

// Very Important to add because it uses
jest.useFakeTimers();

/**
 * 3. Basic example
 */
// Can also be it() or test()
// Describe anatomy of a test
describe('Math Utilities', () => {
  it('should add two numbers correctly', () => {
    const test = 2 + 2;
    expect(test).toEqual(4); // assertion
  });

  it('should multiply two numbers correctly', () => {
    const test = 12 * 12;
    expect(test).toEqual(144); // assertion
  });
});

/**
 * Actual tests
 */
describe('TodoScreen', () => {
  // Create Todo item
  it('should create a Todo item', () => {
    /**
     * 4. 1st Test case Basic example
     */

    const { getByText, getByPlaceholderText } = render(<TodoScreen />);

    const addTodoItemButton = getByText('Add Todo');
    const todoInput = getByPlaceholderText('Add a Todo');

    const newTodoItem = 'Pick up groceries';

    fireEvent.changeText(todoInput, newTodoItem);
    fireEvent.press(addTodoItemButton);

    const createdTodoItem = getByText(newTodoItem);

    expect(createdTodoItem).not.toBeNull();
  });

  // Create Multiple Todo Items
  it('should create multiple Todo items', () => {
    const { getByText, getByPlaceholderText } = render(<TodoScreen />);

    const addTodoItemButton = getByText('Add Todo');
    const todoInput = getByPlaceholderText('Add a Todo');

    const firstTodoItem = 'Pick up groceries';
    const secondTodoItem = 'Create React Native Testing Guide';

    fireEvent.changeText(todoInput, firstTodoItem);
    fireEvent.press(addTodoItemButton);

    fireEvent.changeText(todoInput, secondTodoItem);
    fireEvent.press(addTodoItemButton);

    const firstCreatedTodoItem = getByText(firstTodoItem);
    const secondCreatedTodoItem = getByText(secondTodoItem);

    expect(firstCreatedTodoItem).not.toBeNull();
    expect(secondCreatedTodoItem).not.toBeNull();
  });

  // Delete Todo Item
  it('should delete a Todo item', () => {
    const { getByText, getByPlaceholderText, getByRole, queryByText } = render(
      <TodoScreen />,
    );

    const addTodoItemButton = getByText('Add Todo');
    const todoInput = getByPlaceholderText('Add a Todo');

    const newTodoItem = 'Pick up groceries';

    fireEvent.changeText(todoInput, newTodoItem);
    fireEvent.press(addTodoItemButton);

    const createdTodoItem = getByText(newTodoItem);

    expect(createdTodoItem).not.toBeNull();

    const deleteTodoButton = getByRole('deleteButton');
    fireEvent.press(deleteTodoButton);

    // getBy vs queryBy => getBy... query methods fail (throws error) when there is no matching element (null) but queryBy... methods don’t throw an error when no element (null) is found. We don’t want to get error from the line of fetching element. We want to get the error from the last line of TEST suit that is “expect”. So use queryBy... method instead of getBy....

    // if we use getBy it will error out the test before assertion (expect) because item does not exist in the todo list
    // queryBy does not fail test before assertion (expect)
    const deletedItem = queryByText(newTodoItem);

    expect(deletedItem).toBeNull();
  });

  // Show error message on empty input
  it('should show error message on empty input', () => {
    const { getByText } = render(<TodoScreen />);

    const addTodoItemButton = getByText('Add Todo');

    fireEvent.press(addTodoItemButton);

    const errorMessage = getByText('Please enter a valid todo');

    expect(errorMessage).not.toBeNull();
  });

  // Error message should disappear once a valid item is created
  it('should show remove error message on adding valid input', () => {
    const { getByText, getByPlaceholderText, queryByDisplayValue } = render(
      <TodoScreen />,
    );

    const addTodoItemButton = getByText('Add Todo');

    fireEvent.press(addTodoItemButton);

    const todoInput = getByPlaceholderText('Add a Todo');

    const newTodoItem = 'Pick up groceries';

    fireEvent.changeText(todoInput, newTodoItem);
    fireEvent.press(addTodoItemButton);

    const errorMessage = queryByDisplayValue('Please enter a valid todo');

    expect(errorMessage).toBeNull();
  });

  // Navigation from Todo to Recipes screen

  it('should navigate to Login screen on pressing food icon', () => {
    const navigation = { navigate: jest.fn() };

    const { getByRole } = render(<TodoScreen navigation={navigation} />);

    const recipesLoginButton = getByRole('RecipesLoginButton');
    fireEvent.press(recipesLoginButton);

    expect(navigation.navigate).toHaveBeenCalledWith('Login');
  });
});
