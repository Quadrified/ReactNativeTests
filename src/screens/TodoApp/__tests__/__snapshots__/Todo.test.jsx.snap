// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoScreen renders correctly according to the snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#F6F6F6",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#005A9C",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 24,
        }
      }
    >
      Todo List
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      role="RecipesLoginButton"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#FFFFFF",
              "fontSize": 24,
            },
            undefined,
            {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        î¯²
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "marginHorizontal": 10,
        "marginVertical": 16,
      }
    }
  >
    <View
      style={
        {
          "padding": 5,
        }
      }
    >
      <View
        style={{}}
      >
        <View
          pointerEvents="none"
          style={
            [
              {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 6,
              },
              {
                "top": 0,
              },
              {
                "backgroundColor": "rgba(255, 251, 254, 1)",
                "borderColor": "rgba(121, 116, 126, 1)",
                "borderRadius": 8,
                "borderWidth": 1,
              },
              undefined,
            ]
          }
          testID="text-input-outline"
        />
        <View
          style={
            [
              {
                "flexGrow": 1,
                "paddingBottom": 0,
              },
              {
                "minHeight": 56,
                "paddingTop": 0,
              },
            ]
          }
        >
          <TextInput
            cursorColor="rgba(103, 80, 164, 1)"
            editable={true}
            maxFontSizeMultiplier={1.5}
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onLayout={[Function]}
            onSubmitEditing={[Function]}
            placeholder="Add a Todo"
            placeholderTextColor="rgba(73, 69, 79, 1)"
            selectionColor="rgba(103, 80, 164, 1)"
            style={
              [
                {
                  "flexGrow": 1,
                  "margin": 0,
                },
                {
                  "height": 56,
                },
                {
                  "paddingBottom": 0,
                  "paddingTop": 0,
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "letterSpacing": 0.15,
                  "lineHeight": undefined,
                  "minWidth": 65,
                  "paddingHorizontal": 16,
                  "textAlign": "left",
                  "textAlignVertical": "center",
                },
                false,
                [
                  {},
                ],
                undefined,
              ]
            }
            testID="text-input-outlined"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#005A9C",
            "borderRadius": 8,
            "marginTop": 12,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="AddTodoButton-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "#005A9C",
              "borderColor": "transparent",
              "borderRadius": 8,
              "borderStyle": "solid",
              "borderWidth": 0,
              "flex": undefined,
              "minWidth": 64,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="AddTodoButton-container"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 8,
                },
              ]
            }
            testID="AddTodoButton"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "writingDirection": "ltr",
                    },
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        undefined,
                        {
                          "textTransform": "uppercase",
                        },
                        {
                          "color": "rgba(255, 255, 255, 1)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        {
                          "fontSize": 18,
                        },
                      ],
                    ],
                  ]
                }
                testID="AddTodoButton-text"
              >
                Add Todo
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "marginVertical": 15,
        "paddingHorizontal": 15,
      }
    }
  >
    <RCTScrollView
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      style={
        {
          "width": "100%",
        }
      }
      viewabilityConfigCallbackPairs={[]}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;
